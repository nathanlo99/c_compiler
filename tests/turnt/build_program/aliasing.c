
int wain(int a, int b) {
  int *c = NULL;
  c = &a;

  a = 69; // Seemingly unused assignment
  b = *c; // Sneakily used here
  a = 70;

  return a + b; // 69 + 70 = 139
}
