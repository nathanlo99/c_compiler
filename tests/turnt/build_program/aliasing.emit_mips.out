                                ; // Simplified C code:
                                ; int wain(int a, int b) {
                                ;   int* c = NULL;
                                ;   c = &a;
                                ;   a = 69;
                                ;   b = *c;
                                ;   a = 70;
                                ;   return (a + b);
                                ; }
                                ; 
lis $4
.word 4
slt $11, $0, $4                 ; $11 = ($0 < $4) = 1
beq $0, $0, wain                ; Done prologue, jumping to wain
                                ; 
                                ; Generating code for wain
wain:
sw $1, -4($30)                  ;   push $1
sub $30, $30, $4                ;   ^
sw $2, -4($30)                  ;   push $2
sub $30, $30, $4                ;   ^
sub $29, $30, $4
sw $11, -4($30)                 ;   push $11
sub $30, $30, $4                ; Declaration c
                                ; Code for statements:
                                ; c = &a;
add $3, $4, $4
add $3, $3, $29
sw $3, 0($29)
                                ; a = 69;
lis $3
.word 69                        ; Loading the literal 69
sw $3, 8($29)
                                ; b = *c;
lw $3, 0($29)                   ; Loading c
lw $3, 0($3)
sw $3, 4($29)
                                ; a = 70;
lis $3
.word 70                        ; Loading the literal 70
sw $3, 8($29)
                                ; Code for return value:
lw $3, 8($29)                   ; Loading a
sw $3, -4($30)                  ;   push $3
sub $30, $30, $4                ;   ^
lw $3, 4($29)                   ; Loading b
add $30, $30, $4                ;   pop $5
lw $5, -4($30)                  ;   ^
add $3, $5, $3
jr $31                          ; Done generating code for wain
                                ; Number of assembly instructions: 31
